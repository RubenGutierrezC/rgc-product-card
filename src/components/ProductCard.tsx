import React, { createContext, CSSProperties } from 'react';
import styles from '../styles/styles.module.css';
import useProduct from '../hooks/useProduct';
import { InitialValues, ProductCartHandlers } from '../interface/interface';
import {
  ProductContextProps,
  Product,
  OnChangeArgs,
} from '../interface/interface';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  className?: string;
  style?: CSSProperties;
  onChange?: (arg: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
  children?: (args: ProductCartHandlers) => JSX.Element;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ product, increaseBy, counter, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children &&
          children({
            count: counter,
            maxCount,
            isMaxCountReached,
            reset,
            product,
            increaseBy,
          })}
      </div>
    </Provider>
  );
};
