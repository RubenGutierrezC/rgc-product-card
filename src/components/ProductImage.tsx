import React, { useContext, CSSProperties } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface ProductImageProps {
  image?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({
  image,
  className,
  style,
}: ProductImageProps) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (image) {
    imgToShow = image;
  } else if (product?.image) {
    imgToShow = product?.image;
  } else {
    imgToShow = noImage;
  }

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Product"
      style={style}
    />
  );
};
