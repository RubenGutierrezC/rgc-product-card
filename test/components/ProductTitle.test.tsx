import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('debe mostrar el componente correctamente con el titulo personalizdos', () => {
    const wrapper = renderer
      .create(<ProductTitle title="Custom Product" className="custom-class" />)
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar el component con el nombre del product', () => {
    const wrapper = renderer
      .create(
        <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
