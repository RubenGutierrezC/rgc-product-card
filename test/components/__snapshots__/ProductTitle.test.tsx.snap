// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostrar el component con el nombre del product 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffe Mug - Sin imagen
  </span>
</div>
`;

exports[`ProductTitle debe mostrar el componente correctamente con el titulo personalizdos 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Product
</span>
`;
